### PR TITLE
planner: recalculate as-of ts of staleread when plan is cached

### LINKED ISSUE #43044: prepare stmt with stale read get wrong result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Run playground with the following TiKV config:

```toml
[resolved-ts]
advance-ts-interval = "100ms"
```

Run this case.

```sql
/* init */ drop table if exists t;
/* init */ create table t(id int primary key, v int);

/* init */ insert into t values(1, 1), (2, 2);

/* t1 */ select sleep(2);
/* t1 */ select * from t as of timestamp now() - interval 1 second;
/* t1 */ prepare stmt from 'select * from t as of timestamp now() - interval 1 second';
/* t1 */ execute stmt;

/* t2 */ insert into t values(3, 3);

/* t1 */ select sleep(1);
/* t1 */ select * from t as of timestamp now() - interval 1 second;
/* t1 */ execute stmt;

/* t1 */ select sleep(1);
/* t1 */ select * from t as of timestamp now() - interval 1 second;
/* t1 */ execute stmt;
```

### 2. What did you expect to see? (Required)

select and execute always returns the same result.

### 3. What did you see instead (Required)

<details>

<summary><s>The previous result, there is a issue my test tool, which does not read data from a execute statement</s></summary>

```sql
/* init */ drop table if exists t;
-- init >> 0 rows affected
/* init */ create table t(id int primary key, v int);
-- init >> 0 rows affected
/* init */ insert into t values(1, 1), (2, 2);
-- init >> 2 rows affected
/* t1 */ select sleep(2);
-- t1 >> +----------+
-- t1    | SLEEP(2) |
-- t1    +----------+
-- t1    |        0 |
-- t1    +----------+
/* t1 */ select * from t as of timestamp now() - interval 1 second;
-- t1 >> +----+---+
-- t1    | ID | V |
-- t1    +----+---+
-- t1    |  1 | 1 |
-- t1    |  2 | 2 |
-- t1    +----+---+
/* t1 */ prepare stmt from 'select * from t as of timestamp now() - interval 1 second';
-- t1 >> 0 rows affected
/* t1 */ execute stmt;
-- t1 >> 0 rows affected
/* t2 */ insert into t values(3, 3);
-- t2 >> 1 rows affected
/* t1 */ select sleep(1);
-- t1 >> +----------+
-- t1    | SLEEP(1) |
-- t1    +----------+
-- t1    |        0 |
-- t1    +----------+
/* t1 */ select * from t as of timestamp now() - interval 1 second;
-- t1 >> +----+---+
-- t1    | ID | V |
-- t1    +----+---+
-- t1    |  1 | 1 |
-- t1    |  2 | 2 |
-- t1    +----+---+
/* t1 */ execute stmt;
-- t1 >> 0 rows affected
/* t1 */ select sleep(1);
-- t1 >> +----------+
-- t1    | SLEEP(1) |
-- t1    +----------+
-- t1    |        0 |
-- t1    +----------+
/* t1 */ select * from t as of timestamp now() - interval 1 second;
-- t1 >> +----+---+
-- t1    | ID | V |
-- t1    +----+---+
-- t1    |  1 | 1 |
-- t1    |  2 | 2 |
-- t1    |  3 | 3 |
-- t1    +----+---+
/* t1 */ execute stmt;
-- t1 >> 0 rows affected
```

</details>

~~Note this is also a anomaly. After sleep 1 second, `as of timestamp now() - interval 1 second` should see the previous insert `(3, 3)`.~~

<details>

<summary><s>Case details</s></summary>

```sql
/* t2 */ insert into t values(3, 3);
-- t2 >> 1 rows affected
/* t1 */ select sleep(1);
-- t1 >> +----------+
-- t1    | SLEEP(1) |
-- t1    +----------+
-- t1    |        0 |
-- t1    +----------+
/* t1 */ select * from t as of timestamp now() - interval 1 second;
-- t1 >> +----+---+
-- t1    | ID | V |
-- t1    +----+---+
-- t1    |  1 | 1 |
-- t1    |  2 | 2 |
-- t1    +----+---+
```

</details>

It's caused by `now()` accurate issue.

---

Run this statement with mysql client, got the following output, the resultset of `execute stmt` is not updated after (3, 3) is inserted.

```sql
MySQL [test]> /* init */ create table t(id int primary key, v int);
Query OK, 0 rows affected (0.134 sec)

MySQL [test]> 
MySQL [test]> /* init */ insert into t values(1, 1), (2, 2);
Query OK, 2 rows affected (0.006 sec)
Records: 2  Duplicates: 0  Warnings: 0

MySQL [test]> 
MySQL [test]> /* t1 */ select sleep(2);
+----------+
| sleep(2) |
+----------+
|        0 |
+----------+
1 row in set (2.001 sec)

MySQL [test]> /* t1 */ select * from t as of timestamp now() - interval 1 second;
+----+------+
| id | v    |
+----+------+
|  1 |    1 |
|  2 |    2 |
+----+------+
2 rows in set (0.001 sec)

MySQL [test]> /* t1 */ prepare stmt from 'select * from t as of timestamp now() - interval 1 second';
Query OK, 0 rows affected (0.001 sec)

MySQL [test]> /* t1 */ execute stmt;
+----+------+
| id | v    |
+----+------+
|  1 |    1 |
|  2 |    2 |
+----+------+
2 rows in set (0.001 sec)

MySQL [test]> 
MySQL [test]> /* t2 */ insert into t values(3, 3);
Query OK, 1 row affected (0.004 sec)

MySQL [test]> 
MySQL [test]> /* t1 */ select sleep(1);
+----------+
| sleep(1) |
+----------+
|        0 |
+----------+
1 row in set (1.002 sec)

MySQL [test]> /* t1 */ select * from t as of timestamp now() - interval 1 second;
+----+------+
| id | v    |
+----+------+
|  1 |    1 |
|  2 |    2 |
+----+------+
2 rows in set (0.002 sec)

MySQL [test]> /* t1 */ execute stmt;
+----+------+
| id | v    |
+----+------+
|  1 |    1 |
|  2 |    2 |
+----+------+
2 rows in set (0.001 sec)

MySQL [test]> 
MySQL [test]> /* t1 */ select sleep(1);
+----------+
| sleep(1) |
+----------+
|        0 |
+----------+
1 row in set (1.001 sec)

MySQL [test]> /* t1 */ select * from t as of timestamp now() - interval 1 second;
+----+------+
| id | v    |
+----+------+
|  1 |    1 |
|  2 |    2 |
|  3 |    3 |
+----+------+
3 rows in set (0.002 sec)

MySQL [test]> /* t1 */ execute stmt;
+----+------+
| id | v    |
+----+------+
|  1 |    1 |
|  2 |    2 |
+----+------+
2 rows in set (0.002 sec)
```

### 4. What is your TiDB version? (Required)

```
Release Version: v7.1.0-alpha
Edition: Community
Git Commit Hash: 8c9663941fc633bc57edfcfb3edc5da8533ace57
Git Branch: heads/refs/tags/v7.1.0-alpha
UTC Build Time: 2023-04-12 14:36:43
GoVersion: go1.20.3
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv
```

### PR DESCRIPTION
**Pull request by you06**
### What problem does this PR solve?

Issue Number: close #43044

Problem Summary:

### What is changed and how it works?

Recalculate the snapshot ts for staleread when it's provided by as of statement.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the bug that prepared staleread cannot read following changes.
```

### CODE REVIEW

File: sessiontxn/staleread/processor.go [hunk 1]
```diff
@@ -179,7 +182,7 @@ func (p *staleReadProcessor) OnSelectTable(tn *ast.TableName) error {
 		}
 		return nil
 	}
-	return p.evaluateFromStmtTSOrSysVariable(stmtAsOfTS)
+	return p.evaluateFromStmtTSOrSysVariable(stmtAsOfTS, evaluateTS)
```
Comment (cfzjywxk) [inline sessiontxn/staleread/processor.go:185]: What's the difference between `OnExecutePreparedStmt` and `OnSelectTable`, why do we need to calculate the `TS` in two places? Is it possible to merge them into one?

Comment (you06) [reply sessiontxn/staleread/processor.go:185]: Here we set the ts evaluator which needs to be calculated to staleness processor when preparing, and it'll be returned and set to plan cache.

When invoking `OnExecutePreparedStmt`, it only uses the ts evaluator from plan cache, so it's not necessary to set a evaluator that requires calculation to get a timestamp.

File: sessiontxn/staleread/processor.go [hunk 2]
```diff
@@ -201,7 +204,7 @@ func (p *staleReadProcessor) OnExecutePreparedStmt(preparedTSEvaluator Staleness
 			return err
 		}
 	}
-	return p.evaluateFromStmtTSOrSysVariable(stmtTS)
+	return p.evaluateFromStmtTSOrSysVariable(stmtTS, nil)
```
Comment (cfzjywxk) [inline sessiontxn/staleread/processor.go:207]: We could add some comments about the above logic.

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**